### PR TITLE
Fix ARM64 LEA legalization

### PR DESCRIPTION
**Pull request by Penguinwizzard**
Because we weren't checking the types of things, there were a few ways that the conversion of LEA instructions to ADD instructions on ARM64 could go wrong. This changeset fixes legalization of these instructions to properly handle all of the cases.

### CODE REVIEW

File: lib/Backend/arm64/LegalizeMD.cpp [hunk 1]
```diff
@@ -315,7 +315,41 @@ IR::Instr * LegalizeMD::LegalizeLoad(IR::Instr *instr, uint opndNum, LegalForms
...
+        // If the base and index operands on an LEA are different sizes, grow the smaller one,
+        // matching the larger size but keeping the correct signedness for the type
+        IR::RegOpnd* baseOpnd = indirOpnd->GetBaseOpnd();
+        IR::RegOpnd* indexOpnd = indirOpnd->GetIndexOpnd();
+        if (   baseOpnd != nullptr
+            && indexOpnd != nullptr
+            && baseOpnd->GetSize() != indexOpnd->GetSize()
+            )
+        {
+            if (baseOpnd->GetSize() < indexOpnd->GetSize())
```
Comment (sigatrev) [inline lib/Backend/arm64/LegalizeMD.cpp]: I think we should assert(probably outside of the lambda) that baseOpnd is 64 bit. While it is not unreasonable to have a 32 bit index opnd for an LEA after doing 32 bit math, having a 32 base opnd on a 64 bit system is an error.

File: lib/Backend/arm64/LegalizeMD.cpp [hunk 2]
```diff
@@ -315,7 +315,41 @@ IR::Instr * LegalizeMD::LegalizeLoad(IR::Instr *instr, uint opndNum, LegalForms
...
+                indirOpnd->SetBaseOpnd(movInstr->GetDst()->AsRegOpnd());
+            }
+            else
+            {
+                Assert(indexOpnd->GetSize() < baseOpnd->GetSize());
+                IRType largerType = baseOpnd->GetType();
+                Assert(IRType_IsSignedInt(largerType) || IRType_IsUnsignedInt(largerType));
+                IRType sourceType = indexOpnd->GetType();
+                IRType targetType = IRType_IsSignedInt(sourceType) ? IRType_EnsureSigned(largerType) : IRType_EnsureUnsigned(largerType);
+                IR::Instr* movInstr = Lowerer::InsertMove(indexOpnd->UseWithNewType(targetType, instr->m_func), indexOpnd, instr, false);
```
Comment (sigatrev) [inline lib/Backend/arm64/LegalizeMD.cpp]: The downside of correcting the problem during legalization is the need to insert this extra move. If the size was correct at it's source, we could avoid the extra instruction in any case where we could be sure that the top 32 bits were empty (after any 32 operations were performed). That said, until we have more rigorous testing for arm64 this is probably the better approach.

Comment (Penguinwizzard) [reply lib/Backend/arm64/LegalizeMD.cpp]: In any case, we want this sort of behavior for release builds, even if we change it to assert in debug builds - after a call to legalize, we want to be able to trust that the instructions we have are legal.

File: lib/Backend/arm64/LegalizeMD.cpp (hunk 1 above)
Comment (Penguinwizzard) [reply lib/Backend/arm64/LegalizeMD.cpp]: Good point, added to updated version.

**Pull request merged**